### PR TITLE
Add metrics, conditions to HPA details page

### PR DESCRIPTION
**Pull request by rhamilto**
With an assist from @spadgett and @benjaminapetersen 

screen shot 2018-06-15 at 11 47 36 am

localhost_9000_k8s_ns_robbhamilton_horizontalpodautoscalers_example_

### CODE REVIEW

File: frontend/public/components/hpa.tsx [hunk 1]
```diff
@@ -32,31 +32,186 @@ const menuActions = [
...
+  //   {
+  //     "type": "Resource",
+  //     "resource": {
+  //       "currentAverageValue": "5",
+  //     }
+  //   }
+  // ];
+  // const mockCurrentMetrics = null;
+
+  // if (mockCurrentMetrics) {
```
Comment (benjaminapetersen) [inline frontend/public/components/hpa.tsx]: You might try reducing your `switch` statements down a bit, use `return` to your advantage in the `.map`:

```JavaScript
let currentMetrics = hpa.status.currentMetrics.map((currentMetric, i) => {
  switch(currentMetric.type) {
    case 'External':
      return [currentMetric.external.currentAverageValue, currentMetric.external.currentValue];
    case 'Object':
      return currentMetric.object.currentValue;
    case 'Pods':
      return currentMetric.pods.currentAverageValue;
    case 'Resource':
      return currentMetric.resource.name === 'cpu' ?
      [currentMetric.resource.currentAverageUtilization + '%', currentMetric.resource.currentAverageValue] :
      currentMetric.resource.currentAverageValue;
    default:
      return null;
  }
});
```
You can eliminate quite a few lines.

Comment (benjaminapetersen) [reply frontend/public/components/hpa.tsx]: Might even want to use a `Class` and break some of this out into separate functions...

Comment (benjaminapetersen) [reply frontend/public/components/hpa.tsx]: Heh, sorry but got me thinking, you can even eliminate the switch if you want to:

```JavaScript 

let currentMetrics = hpa.status.currentMetrics.map((currentMetric, i) => {
  return {
    'External' : [currentMetric.external.currentAverageValue, currentMetric.external.currentValue],
    'Object' : currentMetric.object.currentValue,
    'Pods' : currentMetric.pods.currentAverageValue,
    'Resource' : currentMetric.resource.name === 'cpu' ?
      [currentMetric.resource.currentAverageUtilization + '%', currentMetric.resource.currentAverageValue] :
      currentMetric.resource.currentAverageValue
  // the type will match a key above
  }[currentMetric.type] || null;
});
```
fewer lines + pretty clear key-value match.

File: frontend/public/components/hpa.tsx [hunk 2]
```diff
@@ -31,8 +35,98 @@ const menuActions = [
...
+const MetricsTable: React.SFC<MetricsTableProps> = ({obj: hpa}) => {
+  const namespace = hpa.metadata.namespace;
+  // https://github.com/kubernetes/api/blob/master/autoscaling/v2beta1/types.go
+  const metrics = hpa.spec.metrics.map((metric, i) => {
+    let typeLabel, targetValue, currentValue;
+    const currentMetrics = _.get(hpa, ['status', 'currentMetrics', i]);
+    switch (metric.type) {
+      case 'External':
+        typeLabel = metric.external.metricName;
+        // TODO: show metric selector for external metrics?
```
Comment (benjaminapetersen) [inline frontend/public/components/hpa.tsx]: I would suggest trying to avoid nested `if` statements within a `switch`.  It's hard to follow the flow & pretty easy to break in later updates.  The  valueless`let typeLabel, targetValue, currentValue;` at the top is a pretty good indicator we can break it down into smaller bits.

Comment (benjaminapetersen) [reply frontend/public/components/hpa.tsx]: See my previous comment, I'd even wager you can avoid the `switch` altogether to improve readability.

**Comment by benjaminapetersen** [comment]
@rhamilto @spadgett I'd suggest breaking down into soemthing like this, but I'll defer to y'all:

```JavaScript

const theHTML = (typeLabel, currentValue, targetValue, key) => <div className="row" key={key}>
  <div className="col-xs-6">
    {typeLabel}
  </div>
  <div className="col-xs-3">
    {currentValue || '-'}
  </div>
  <div className="col-xs-3">
    {targetValue || '-'}
  </div>
</div>;


const makesFooBarBaz = (currentMetrics, i) => theHTML(
  `${metric.pods.metricName} on pods`, 
  _.get(currentMetrics, 'pods.currentAverageValue'), 
  metric.pods.targetAverageValue,
  i
);


const MetricsTable: React.SFC<MetricsTableProps> = ({obj: hpa}) => {
  switch(thing) {
    case 'Resource':
      return makesFooBarBaz(metric, currentMetrics);
      break;
    case 'Foo':
      return makeOtherThing();
      break;
    case 'Bar':
      return makeOtherThing();
      break; 
  }
};
```

File: frontend/public/components/hpa.tsx [hunk 3]
```diff
@@ -31,8 +35,98 @@ const menuActions = [
...
+          typeLabel += ' (as a percentage of request)';
+          const utilization = _.get(currentMetrics, 'resource.currentAverageUtilization');
+          if (_.isNumber(utilization)) {
+            currentValue = `${utilization}%`;
+          }
+          targetValue = `${metric.resource.targetAverageUtilization}%`;
+        } else if (metric.resource.targetAverageValue) {
+          const currentValueRaw = _.get(currentMetrics, 'resource.currentAverageValue');
+          currentValue = humanizeResourceValue(currentValueRaw, metric.resource.name);
+          targetValue = humanizeResourceValue(metric.resource.targetAverageValue, metric.resource.name);
```
Comment (spadgett) [inline frontend/public/components/hpa.tsx]: It looks like our use of humanize is wrong. Debugging now.

Comment (spadgett) [reply frontend/public/components/hpa.tsx]: `humanizeMem` assumes you're passing the value in bytes instead of a string like `50Mi`. I'd rather not do that since the value in spec is likely using the units we want to show (e.g., we shouldn't convert 50MB to MiB and not have a round number).

I think it might be easiest for the moment to remove the humanize for current value as well since otherwise they will have different formatting (current: 30.22 MiB, target: 50Mi). We can always revisit this.

**Comment by spadgett** [comment]
@benjaminapetersen Updated, mind taking another look?

**Comment by rhamilto** [comment]
Merged @spadgett's PR.  @benjaminapetersen, PTAL.

File: frontend/public/components/hpa.tsx [hunk 4]
```diff
@@ -31,8 +34,113 @@ const menuActions = [
...
+  </div>
+  <div className="col-xs-3">
+    {target || '-'}
+  </div>
+</div>;
+
+const externalRow = (metric, current, key) => {
+  const { external } = metric;
+  const type = external.metricName;
+
```
Comment (benjaminapetersen) [inline frontend/public/components/hpa.tsx]: Thoughts on splitting these apart?

```JavaScript 
// vs 
const currentValue = external.targetAverageValue
  ? _.get(current, 'object.currentAverageValue') 
  : _.get(current, 'object.currentValue');

const targetValue = external.targetAverageValue
  ? external.targetAverageValue
  : external.targetValue;
```
is a nit, but a little more declarative (I'm using the `ternary` the way I see it used elsewhere in the codebase).

File: frontend/public/components/hpa.tsx [hunk 5]
```diff
@@ -31,8 +34,113 @@ const menuActions = [
...
+  </React.Fragment>;
+  const currentValue = _.get(current, 'object.currentValue');
+  const targetValue = object.targetValue;
+
+  return <MetricsRow key={key} type={type} current={currentValue} target={targetValue} />;
+};
+
+const podRow = (metric, current, key) => {
+  const { pods } = metric;
+  const type = `${pods.metricName} on pods`;
```
Comment (benjaminapetersen) [inline frontend/public/components/hpa.tsx]: This instance of `pods` does not need to be linked (to pod list page) correct? We are just linking to specific resources, like [line 70](https://github.com/openshift/console/pull/97/files#diff-c721d4405104cb163ef8cda009eb4bf1R70).

File: frontend/public/components/hpa.tsx [hunk 6]
```diff
@@ -31,8 +34,113 @@ const menuActions = [
...
+  const currentValue = _.get(current, 'pods.currentAverageValue');
+  const targetValue = pods.targetAverageValue;
+
+  return <MetricsRow key={key} type={type} current={currentValue} target={targetValue} />;
+};
+
+const resourceRow = (metric, current, key) => {
```
Comment (benjaminapetersen) [inline frontend/public/components/hpa.tsx]: Wondering if this can be more declarative as well, if we split each bit out && use `const`:

```JavaScript 
const getUtilization = (current) => {
  let val = _.get(current, 'resource.currentAverageUtilization'); 
  return _.isNumber(val) ? `${val}%` : val;
}

const resourceRow = (metric, current, key) => {
  const currentValue = getUtilization(current);
  const targetValue = metric.resource.targetAverageUtilization
    ? `${metric.resource.targetAverageUtilization}%` 
    :  metric.resource.targetAverageValue;

  const type = metric.resource.targetAverageUtilization 
    ? <React.Fragment>resource {metric.resource.name}</React.Fragment> 
    : <React.Fragment>
        {type}&nbsp;<span className="small text-muted">(as a percentage of request)</span>
      </React.Fragment>;

  return <MetricsRow key={key} type={type} current={currentValue} target={targetValue} />;
};
```
Not tested, very possible I missed something.

File: frontend/public/components/hpa.tsx [hunk 7]
```diff
@@ -31,8 +34,113 @@ const menuActions = [
...
+      </div>
+      <div className="co-m-table-grid__body">
+        {hpa.spec.metrics.map((metric, i) => {
+          // https://github.com/kubernetes/api/blob/master/autoscaling/v2beta1/types.go
+          const current = _.get(hpa, ['status', 'currentMetrics', i]);
+          switch (metric.type) {
+            case 'External':
+              return externalRow(metric, current, i);
+            case 'Object':
+              return objectRow(metric, current, i, hpa.metadata.namespace);
```
Comment (benjaminapetersen) [inline frontend/public/components/hpa.tsx]: Any reason `key` is not the last arg for just this one?

Comment (benjaminapetersen) [reply frontend/public/components/hpa.tsx]: Passing `key` to `fn`s seems fine, I think.  Had a short chat w/@spadgett about it.

File: frontend/public/components/hpa.tsx (hunk 5 above)
Comment (spadgett) [reply frontend/public/components/hpa.tsx]: Yeah we can't link directly to the pods because we don't know what label selector to use to filter the list. It's whatever pods are owned by the thing being scaled.

File: frontend/public/components/hpa.tsx [hunk 8]
```diff
@@ -130,4 +244,14 @@ export type HorizontalPodAutoscalersPageProps = {
 export type HorizontalPodAutoscalersDetailsPageProps = {
   match: any,
 };
+
+export type MetricsTableProps = {
+  obj: any,
+};
+
+export type MetricsRowProps = {
```
[the remark this answers is not in the document]
Comment (benjaminapetersen) [reply frontend/public/components/hpa.tsx]: I think the compiler currently complains if you don't export them.  I tried not to with the pipelines PR, it gets grumpy.

File: frontend/public/components/hpa.tsx (hunk 7 above)
Comment (spadgett) [reply frontend/public/components/hpa.tsx]: Yeah I think putting key last always probably makes sense

File: frontend/public/components/hpa.tsx (hunk 5 above)
Comment (benjaminapetersen) [reply frontend/public/components/hpa.tsx]: Ok great.

File: frontend/public/components/hpa.tsx (hunk 6 above)
Comment (rhamilto) [reply frontend/public/components/hpa.tsx]: I don't think this is parallel to the existing.

Comment (benjaminapetersen) [reply frontend/public/components/hpa.tsx]: Its very close, prob not perfect as I didn't run it to test.  The point is, its possible to isolate each variable & set it clearly.  Testing `metric.resource.targetAverageUtilization` twice to isolate the values makes it much easier to understand now && change in the future.

Comment (benjaminapetersen) [reply frontend/public/components/hpa.tsx]: Can inline the `getUtilization()` as well if preferred:

```JavaScript
const utilization = _.get(current, 'resource.currentAverageUtilization'); 
const currentValue = _.isNumber(utilization) ? `${utilization}%` : utilization;
```
Commenting more on isolating each variable than an exact impl.

Comment (rhamilto) [reply frontend/public/components/hpa.tsx]: I inlined getUtilization and added currentAverageValue in parens as that's what the CLI does.

screen shot 2018-06-01 at 10 47 13 am

Comment (benjaminapetersen) [reply frontend/public/components/hpa.tsx]: That works, though if it *might* be reusable later, it's much better as a function.  I tend to not want to leave it for someone else to refactor later if they find that they need it.  That leads us to a lot of duplication.

I'm still pretty concerned about the complexity of the `if` statement.  There are many things happening in parallel that can be extracted.

Comment (rhamilto) [reply frontend/public/components/hpa.tsx]: @spadgett, your thoughts?

Comment (benjaminapetersen) [reply frontend/public/components/hpa.tsx]: @rhamilto describe again quick the tricky part of this -- which vars may or may not be set together.  That might be a good clue.  I don't think its explicit in this current implementation, but maybe can be made explicit.

It sounded confusing, and I think my suggested refactor missed it, which probably confirms that there is some confusing logic masked here.

Comment (benjaminapetersen) [reply frontend/public/components/hpa.tsx]: Curious what the `_.isNumber(utilization)` really means.  There is something specific going on there, I think it would be helpful to name it (named fn);

The more we look at it and talk about it, the more I want a little more clarity on some of the bits going on.

Comment (spadgett) [reply frontend/public/components/hpa.tsx]: We're just checking if there's a value. 0 should evaluate to true even though it's falsey, so we use `_.isNumber`.

```
_.isNumber(undefined)
false
_.isNumber(null)
false
_.isNumber(0)
true
```

Comment (benjaminapetersen) [reply frontend/public/components/hpa.tsx]: If its a number, its assumed to be a `%`, otherwise `utililzation` will include its own suffix, indicating the kind of metric it is?

Comment (benjaminapetersen) [reply frontend/public/components/hpa.tsx]: In the end, I just don't think these need to be intertwined.  95% confident they can still be `const`, can pivot on `metric.resource.targetAverageUtilization` separately & be set cleanly w/o a ton of effort.  It would be more declarative, rely less on side effects & look more consistent w/our other react code / react code in general.  I'd rather set our trajectory away from intertwined blocks.

That said, @spadgett @rhamilto if you strongly disagree, feel free to move on.

Comment (rhamilto) [reply frontend/public/components/hpa.tsx]: @benjaminapetersen worked together to come up with the following.  @spadgett, how about something like this?

```javascript
const getUtilization = (metric, current, metricUtilization) => {
  const currentAverageValue = _.get(current, 'resource.currentAverageValue');
  const utilization = _.get(current, 'resource.currentAverageUtilization');
  const currentValueUtilization = _.isNumber(utilization)
    ? `${utilization}% (${currentAverageValue})`
    : currentAverageValue;

  return metricUtilization
    ? currentValueUtilization
    : currentAverageValue;
}

const resourceRow = (metric, current, key) => {
  const metricUtilization = metric.resource.targetAverageUtilization;
  const currentValue = getUtilization(metric, current, metricUtilization);
  const targetValue = metricUtilization
    ? `${metricUtilization}%`
    :  metric.resource.targetAverageValue;
  const resource = `resource ${metric.resource.name}`;

  const type = metricUtilization
    ? <React.Fragment>{resource}</React.Fragment>
    : <React.Fragment>
        {resource}&nbsp;<span className="small text-muted">(as a percentage of request)</span>
      </React.Fragment>;

  return <MetricsRow key={key} type={type} current={currentValue} target={targetValue} />;
};
```

**Comment by rhamilto** [comment]
Revised to included changes based on @benjaminapetersen's feedback.  Ready for final review.

**Comment by benjaminapetersen** [comment]
LGTM though there is a GUI Test error, I see a lot of this repeated:

      2018/06/15 13:35:16 CheckOrigin: Proxy has no configured Origin. Allowing origin [http://localhost:9000] to wss://teamui-api.console.team.coreos.systems:443/api/v1/namespaces?watch=true&resourceVersion=35119916

**Comment by benjaminapetersen** [comment]
Seems like an infra issue? or a flake?

**Comment by rhamilto** [comment]
> Seems like an infra issue? or a flake?

I'm seeing that locally when connecting to free-int.  @spadgett?

**Comment by spadgett** [comment]
jenkins retest

**Comment by spadgett** [comment]
> 1) Kubernetes resource CRUD operations HorizontalPodAutoscaler : displays detail view for new resource instance
   Error: Timeout - Async callback was not invoked within timeout specified by jasmine.DEFAULT_TIMEOUT_INTERVAL.

I think this is probably a legit failure. @rhamilto can you take a look? You should be able to use

```
$ yarn run test-gui
```

File: frontend/public/components/hpa.tsx [hunk 9]
```diff
@@ -31,8 +34,118 @@ const menuActions = [
...
+};
+
+const resourceRow = (metric, current, key) => {
+  const metricUtilization = metric.resource.targetAverageUtilization;
+  const resource = `resource ${metric.resource.name}`;
+  const type = metricUtilization
+    ? <React.Fragment>{resource}&nbsp;<span className="small text-muted">(as a percentage of request)</span></React.Fragment>
+    : <React.Fragment>
+      {resource}
+    </React.Fragment>;
```
Comment (spadgett) [inline frontend/public/components/hpa.tsx]: React.Fragement isn't needed here.

```jsx
const type = metricUtilization
  ? <React.Fragment>{resource}&nbsp;<span className="small text-muted">(as a percentage of request)</span></React.Fragment>
  : resource;
```

File: frontend/public/components/hpa.tsx [hunk 10]
```diff
@@ -31,8 +34,118 @@ const menuActions = [
...
+const podRow = (metric, current, key) => {
+  const { pods } = metric;
+  const type = `${pods.metricName} on pods`;
+  const currentValue = _.get(current, 'pods.currentAverageValue');
+  const targetValue = pods.targetAverageValue;
+
+  return <MetricsRow key={key} type={type} current={currentValue} target={targetValue} />;
+};
+
+const getUtilization = (current, metricUtilization) => {
```
Comment (spadgett) [inline frontend/public/components/hpa.tsx]: This name is confusing:

* It's not clear if it gets target average utilization or current average utilization
* This function sometimes returns current average value instead of a utilization percentage, so it's not always returning "utilization"
* It's not clear that this is only for "Resource" type metrics

I would suggest calling it something like `getCurrentResourceUsage` as written.

But I think this would be better:

```jsx
const getCurrentResourceUtilization = currentMetric => {
  const currentUtilization = _.get(current, 'resource.currentAverageUtilization');
  const currentAverageValue = _.get(current, 'resource.currentAverageValue');
  // Fallback to currentAverageValue if utilization is not set.
  return _.isNumber(currentUtilization) ? `${currentUtilization}% (${currentAverageValue})`  : currentAverageValue;
};
```

Then only call it below when we want utilization:

```js
const currentValue = metricUtilization ? getUtilization(current) : _.get(current, 'resource.currentAverageValue');
const targetValue = metricUtilization ? `${metricUtilization}%` : metric.resource.targetAverageValue;
```

If we're going to check the same condition twice, it's better if the checks are side-by-side so that they don't get out of sync. The proposed refactor puts them in separate methods.

If we refactor code to pull out methods, it's important that the helpers do one clear thing and have good names.
